### DISCUSSION TITLE
Introducing the Markdoc language server and Visual Studio Code extension

### DISCUSSION THREAD

**Discussion by rpaul-stripe** [markdoc/markdoc#407] [Announcements]
The official [Markdoc language server](https://github.com/markdoc/language-server) and [Visual Studio Code extension](https://marketplace.visualstudio.com/items?itemName=Stripe.markdoc-language-support) are now available. The extension enables support for syntax highlighting, validation, autocompletion, formatting, inline partial editing, and a range of other features.

Follow the instructions in the [configuration quickstart](https://github.com/markdoc/language-server#configuration-quickstart) to enable Markdoc for your content and your own custom Markdoc schema. This is important for features like validation and tag auto-completion that rely on information provided by the schema. In addition to providing a configuration that points to the content and schema, you also need to either use a [supported file extension](https://github.com/markdoc/language-server#file-extensions) or configure your workspace to recognize the file extension of your choice as Markdoc content.

The underlying language server that powers the extension is open source and extensible. We published the language server as a [package on npm](https://www.npmjs.com/package/@markdoc/language-server) so that it can be consumed as a dependency and used to make custom variants that are tailored to specific projects or publishing stacks. This extensibility model is still experimental — the public APIs exposed by the language server package are subject to backwards-incompatible changes while we work towards a stable 1.0 release.

It took some time to work through edge cases and land on the right design, but we're excited to share this with the community.

**Comment by romanr** [comment]
Been using VS Code for years and this is the first time I encounter extension that says user have to create some configuration file, and associations file, which seems other extensions do automatically?

**Reply by rpaul-stripe** [reply]
Markdoc tools need to have access to the schema associated with your content in order to provide most language features besides syntax highlighting. The purpose of the configuration file is mainly to point the language server to your schema. It is possible to use the extension without a configuration file if you just want syntax highlighting and don't need other features like validation and auto-completion. This is similar to how VSCode's built-in TypeScript functionality requires a `tsconfig.json` file for many features or the way that the rust-analyzer extension needs a `Cargo.toml` file to be able to resolve dependencies.

The Markdoc extension includes automatic file associations for the following suffixes: `.markdoc`, `.markdoc.md`, or `.mdoc`. If you use one of those suffixes then you don't need to configure a file association—you only need to manually configure a file association if you want to use a different suffix or want to use `.md` for Markdoc instead of Markdown. We don't support `.md` for Markdoc files by default because it creates conflicts with VSCode's built-in Markdown language server. Using a distinct file suffix for Markdoc content ensures that it isn't mistaken for Markdown content.

**Comment by mskri** [comment]
Been trying to get the extension to work with TypeScript. Is it possible to have the schema in `.ts` file? 

For example

```
// src/markdoc/config.ts
import type { Config } from '@markdoc/markdoc';

export const markdocConfig: Config = {
  ...
]
```

```json
...
"schema": {
    "path": "src/markdoc/config.ts",
    "type": "esm",
    "property": "markdocConfig",
    "watch": true
}
...
```

**Reply by rpaul-stripe** [reply]
This is not currently possible. The language server dynamically loads the schema in a running Node.js process, so there's no TypeScript compiler present. We recommend having the config point to a bundled JavaScript build of the schema.

We could consider adding support in the config file for specifying extra command line arguments that should be passed to Node.js when executing the language server, which would make it possible to use custom TypeScript loaders like [tsx](https://github.com/esbuild-kit/tsx).

**Comment by JamesTaxd** [comment]
Does the schema path need to direct to a schema.js file or will a directory with nodes.js and tags.js work or like the Next.js starter repo?
At the moment nothing seems to be happening (barring highlighting) when using it with a directory path.

**Reply by rpaul-stripe** [reply]
It needs to point directly to a file, it can't be a directory.

**Reply by JamesTaxd** [reply]
I don't know if I'm just completely missing it on the docs and example repositories but I can't find an example of a schema.js for Next.js. They all seem to use the folder structure. Any chance I could get linked an example please?

**Reply by mfix-stripe** [reply]
@JamesTaxd you can create a folder structure like this:
```sh
.
└── markdoc/
    ├── tags.js
    ├── nodes.js
    └── config.js
```
where you export the `nodes` and `tags` object from `markdoc/config.js`

and point your VS Code extension at `markdoc/config.js`. That should work!

**Reply by mfix-stripe** [reply]
This pattern is documented here: https://markdoc.dev/docs/nextjs#advanced

**Reply by JamesTaxd** [reply]
Thanks for the help @mfix-stripe, I thought it would be like that but adding that file just prevents the custom nodes from generating at all. Clearly doing something wrong haha

**Comment by NicHaley** [comment]
Got up and running with it today -- very neat :)
